### PR TITLE
Add SaveStartupConfig for SaveOnSet feature

### PR DESCRIPTION
**Pull request by rlucus**
https://github.com/sonic-net/SONiC/issues/1239

### CODE REVIEW

File: translib/transformer/cfg_mgmt_xfmr.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"github.com/Azure/sonic-mgmt-common/translib/tlerr"
+)
+
+// SaveStartupConfig initiates the operation of saving the current content of
+// the ConfigDB to a file that is then used to populate the database during the startup.
+func SaveStartupConfig() error {
+	r := HostQuery("cfg_mgmt.save", []string{})
+	if r.Err != nil {
+		return tlerr.New(Format: "internal SONiC Hostservice communication failure: %w", r.Err.Error())
```
Comment (tomek-US) [inline translib/transformer/cfg_mgmt_xfmr.go:12]: It should be:

```
return tlerr.New("internal SONiC Hostservice communication failure: %w", r.Err.Error())
```

Comment (rlucus) [reply translib/transformer/cfg_mgmt_xfmr.go]: Done

File: translib/transformer/cfg_mgmt_xfmr.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"github.com/Azure/sonic-mgmt-common/translib/tlerr"
+)
+
+// SaveStartupConfig initiates the operation of saving the current content of
+// the ConfigDB to a file that is then used to populate the database during the startup.
+func SaveStartupConfig() error {
+	r := HostQuery("cfg_mgmt.save", []string{})
+	if r.Err != nil {
+		return tlerr.New("internal SONiC Hostservice communication failure: %w", r.Err.Error())
```
Comment (sachinholla) [inline translib/transformer/cfg_mgmt_xfmr.go:12]: IMO `%w` should not be used here.. It is only recognized by `fmt.Errorf` and not by other Sprintf like functions.

**Review by sachinholla — CHANGES_REQUESTED** [review]

File: translib/transformer/cfg_mgmt_xfmr.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package transformer
+
+import (
+	"github.com/Azure/sonic-mgmt-common/translib/tlerr"
+)
+
+// SaveStartupConfig initiates the operation of saving the current content of
+// the ConfigDB to a file that is then used to populate the database during the startup.
+func SaveStartupConfig() error {
```
[the remark this answers is not in the document]
Comment (rlucus) [reply translib/transformer/cfg_mgmt_xfmr.go]: Done

File: translib/transformer/cfg_mgmt_xfmr.go (hunk 2 above)
Comment (rlucus) [reply translib/transformer/cfg_mgmt_xfmr.go]: Done

**Review by sachinholla — APPROVED** [review]

**Review by tomek-US — APPROVED** [review]
Approving based on Sashin's approval.

**Comment by rlucus** [comment]
Updated CL with agreed upon method of using minimum version of rpc_copy

**Comment by rlucus** [comment]
> A question - Say, config changes are made from other apps, but 'config save' not issued yet on the device. Meanwhile, enabling save-on-set does forcefully save all the changes including one of the last gnmi-set operation?

Yes that would be the case.

**Comment by mbalachandar** [comment]
Its out-of-date with base branch, please merge the latest from sonic-net:master

**Comment by rlucus** [comment]
Done

**Review by sachinholla — CHANGES_REQUESTED** [review]
Latest commit is introducing many rpcs and options that are not really supported. This is not a clean approach. Approach of having a translib API would have been better -- copy r s is a standard operation in NOS world.

File: translib/transformer/xfmr_file_mgmt.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		glog.Error("Copy input not provided.")
+		err = errors.New("Input parameters missing.")
+	} else {
+
+		if operand.Input.Source != "running-configuration" || operand.Input.Destination != "startup-configuration" {
+			return nil, errors.New("rpc_copy_cb: Only supports running-configuration -> startup-configuration")
+		}
+
+		glog.Infof("Invoke cfg_mgmt.save %v", options)
+		query_result = HostQuery("cfg_mgmt.save", options)
```
Comment (sachinholla) [inline translib/transformer/xfmr_file_mgmt.go:80]: I don't think there is any host service module named "cfg_mgmt" in the community branch. Are these changes tested?

**Comment by anand-kumar-subramanian** [comment]
> Latest commit is introducing many rpcs and options that are not really supported. This is not a clean approach. Approach of having a translib API would have been better -- copy r s is a standard operation in NOS world.

Anand>> Please have only the necessary RPC implemented end to end. Others will come in when Dell plans to bring it in.

File: translib/transformer/xfmr_file_mgmt.go (hunk 4 above)
Comment (rlucus) [reply translib/transformer/xfmr_file_mgmt.go:80]: There isn't cfg_mgmt in the community branch, but that is beyond the scope of this feature.

**Comment by rlucus** [comment]
> > Latest commit is introducing many rpcs and options that are not really supported. This is not a clean approach. Approach of having a translib API would have been better -- copy r s is a standard operation in NOS world.
> 
> Anand>> Please have only the necessary RPC implemented end to end. Others will come in when Dell plans to bring it in.

They are returning Unimplemented. Do you want them removed from the yang and annotation as well? Please be more specific, it seems as if Sachin is rejecting the whole concept of the compromise we all agreed on.

File: translib/transformer/xfmr_file_mgmt.go (hunk 4 above)
Comment (anand-kumar-subramanian) [reply translib/transformer/xfmr_file_mgmt.go:80]: Is this feature even working without this?

**Comment by anand-kumar-subramanian** [comment]
> > > Latest commit is introducing many rpcs and options that are not really supported. This is not a clean approach. Approach of having a translib API would have been better -- copy r s is a standard operation in NOS world.
> > 
> > 
> > Anand>> Please have only the necessary RPC implemented end to end. Others will come in when Dell plans to bring it in.
> 
> They are returning Unimplemented. Do you want them removed from the yang and annotation as well? Please be more specific, it seems as if Sachin is rejecting the whole concept of the compromise we all agreed on.

Anand>>I meant only the RPC to perform config save needs to be implemented end to end all others will be implemented when Dell plans to bring it in. Other RPCs should not be in the YANG, annotation or the transformer implementation.

File: translib/transformer/xfmr_file_mgmt.go (hunk 4 above)
Comment (anand-kumar-subramanian) [reply translib/transformer/xfmr_file_mgmt.go:77]: @rlucus Can you please create an issue to have the host query working for cfg_mgt and assign it to the relevant team and have the issue as part of this description. Also can say that the feature is not complete with this issue.

**Comment by rlucus** [comment]
> Anand>>I meant only the RPC to perform config save needs to be implemented end to end all others will be implemented when Dell plans to bring it in. Other RPCs should not be in the YANG, annotation or the transformer implementation.

Removed the requested stubs. I believe we're good to go now.

File: models/yang/lint_ignore.ocstyle [hunk 5]
```diff
@@ -0,0 +1 @@
+openconfig-file-mgmt-private.yang
```
Comment (sachinholla) [inline models/yang/lint_ignore.ocstyle:1]: Linter is rejecting this lint_ignore configuration; hence the build error. I think it accepts only the yang module names here and not yang file names. Change to "openconfig-file-mgmt-private".

Comment (rlucus) [reply models/yang/lint_ignore.ocstyle]: This was part of it, but the doctree.py pyang plugin was not up to date with the private repo and was also causing an issue.

**Review by anand-kumar-subramanian — APPROVED** [review]

File: translib/transformer/xfmr_file_mgmt.go (hunk 4 above)
Comment (sachinholla) [reply translib/transformer/xfmr_file_mgmt.go:77]: @rlucus I see a host module function to run config save -- https://github.com/sonic-net/sonic-host-services/blob/6fce4781da508709f11be595e3c7048ffd3138cb/host_modules/config_engine.py#L32

**Comment by rlucus** [comment]
Closing this PR as its no longer needed due to using sonic services client in https://github.com/sonic-net/sonic-gnmi/pull/108

**Pull request closed without merging**